Applications/SimpleShot/layers/centorid_nearest_neighbors.cpp

```diff
@@ -0,0 +1,157 @@
+    ml_loge("centroid nearest layer needs to define num_class for now, please "
+            "check");
+    return ML_ERROR_INVALID_PARAMETER;
+  }
+
+  output_dim[0] = nntrainer::TensorDim({num_class});
+
+  /// weight is a distance map that contains centroid of features of each class
+  auto map_dim =
+    nntrainer::TensorDim({num_class, input_dim[0].getFeatureLen()});
+
+  /// samples seen for the current run to calculate the centroid
+  auto samples_seen = nntrainer::TensorDim({num_class});
+
+  if (weights.empty()) {
+    weights.reserve(1);
+    weights.emplace_back(map_dim, nntrainer::WeightInitializer::WEIGHT_ZEROS,
+                         false, "centroidNN:map");
+    weights.emplace_back(samples_seen,
+                         nntrainer::WeightInitializer::WEIGHT_ZEROS, false,
+                         "centroidNN:num_samples");
+    manager.trackWeights(weights);
+  } else {
+    weights[0].reset(map_dim, nntrainer::WeightInitializer::WEIGHT_ZEROS,
+                     false);
+    weights[1].reset(samples_seen, nntrainer::WeightInitializer::WEIGHT_ZEROS,
+                     false);
+  }
+
+  return ML_ERROR_NONE;
+}
+
+void CentroidNearestNeighborsLayer::forwarding(bool training) {
+  auto &hidden_ = net_hidden[0]->getVariableRef();
+  auto &input_ = net_input[0]->getVariableRef();
+  auto &label = net_hidden[0]->getGradientRef();
+
+  auto &map = weightAt(0).getVariableRef();
+  auto &num_samples = weightAt(1).getVariableRef();
+  auto feature_len = input_dim[0].getFeatureLen();
+
+  auto get_distance = [](const nntrainer::Tensor &a,
+                         const nntrainer::Tensor &b) {
+    return -a.subtract(b).l2norm();
+  };
+
+  if (training) {
+    auto ans = label.argmax();
+
+    for (unsigned int b = 0; b < input_.batch(); ++b) {
+      auto saved_feature = nntrainer::Tensor::Map(map.getData(), {feature_len},
+                                                  ans[b] * feature_len);
+      auto num_sample = num_samples.getValue(0, 0, 0, ans[b]);
+      auto current_feature = input_.getBatchSlice(b, 1);
+      saved_feature.multiply_i(num_sample);
+      saved_feature.add_i(current_feature);
+      saved_feature.divide_i(num_sample + 1);
+      num_samples.setValue(0, 0, 0, ans[b], num_sample + 1);
+    }
+  }
+
+  for (unsigned int i = 0; i < num_class; ++i) {
+    auto saved_feature =
+      nntrainer::Tensor::Map(map.getData(), {feature_len}, i * feature_len);
+
+    auto num_sample = num_samples.getValue(0, 0, 0, i);
+
+    for (unsigned int b = 0; b < input_.batch(); ++b) {
+      auto current_feature = input_.getBatchSlice(b, 1);
+
+      if (num_sample == 0) {
+        hidden_.setValue(b, 0, 0, i, std::numeric_limits<float>::min());
+      } else {
+        hidden_.setValue(b, 0, 0, i,
+                         get_distance(current_feature, saved_feature));
+      }
+    }
+  }
+}
+
+void CentroidNearestNeighborsLayer::calcDerivative() {
+  // do nothing but pass back as this is inplacable operation
```

Comment:
Either this must set derivatives or throw.
IMO, this must throw.

Reply:
Input and output derivatives have different memories, so this must copy derivatives/share memory if needed.
As the comment says to simply pass back derivative, the derivative of this operation is unity?

Reply:
That was previous behavior,
this layer has changed to throw in this case to keep consistency with other layers.